modules/dnn/src/layers/scatterND_layer.cpp

```diff
@@ -0,0 +1,263 @@
+                     ndims, shape, step, data.ptr<char>(),
+                     ind_ndims, ind_shape, ind_step, indices.ptr<char>(),
+                     upd_ndims, upd_shape, upd_step, updates.ptr<char>(),
+                     out.ptr<char>());
+    }
+
+    template<typename T, typename Functor>
+    void forward_impl(const Functor& rd,
+                      int ndims, const int* shape, const size_t* step, const char* input,
+                      int ind_ndims, const int* ind_shape, const size_t* ind_step, const char* indices,
+                      int upd_ndims, const int* upd_shape, const size_t* upd_step, const char* updates,
+                      char* out)
+    {
+        size_t input_total = 1;
+        for (int i = 0; i < ndims; i++)
+            input_total *= shape[i];
+        std::memcpy(out, input, input_total * sizeof(T));
+
+        int k = ind_shape[ind_ndims - 1]; // last dim of indices
+        size_t total_div_k = 1;
+        for (int i = 0; i < ind_ndims - 1; i++)
+            total_div_k *= ind_shape[i];
+
+        size_t updates_size = 1;
+        for (int i = ind_ndims - 1; i < upd_ndims; i++)
+            updates_size *= upd_shape[i];
+
+        size_t inp_start_offset = 0;
+        size_t ind_start_offset = 0;
+        size_t upd_start_offset = 0;
+        for (size_t i = 0; i < total_div_k; i++)
+        {
+            size_t t = i;
+            for (int j = ind_ndims - 2; j >= 0; j--)
+            {
+                size_t idx = t / ind_shape[j];
+                int offset_at_idx = (int)(t - idx * ind_shape[j]);
+                update_indices[j] = offset_at_idx;
+                t = idx;
+            }
+
+            ind_start_offset = 0;
+            for (int j = 0; j < ind_ndims - 1; j++)
+                ind_start_offset += update_indices[j] * ind_step[j];
+
+            const T* p_indices = (const T*)(indices + ind_start_offset);
+            std::fill(inp_current_indices.begin(), inp_current_indices.end(), 0);
+            for (int j = 0; j < k; j++)
+            {
+                inp_current_indices[j] = p_indices[j];
+            }
+
+            inp_start_offset = 0;
+            for (int j = 0; j < ndims; j++)
+            {
+                inp_start_offset += inp_current_indices[j] * step[j];
+            }
+
+            std::fill(upd_current_indices.begin(), upd_current_indices.end(), 0);
+            for (int j = 0; j < ind_ndims - 1; j++)
+            {
+                upd_current_indices[j] = update_indices[j];
+            }
+
+            upd_start_offset = 0;
+            for (int j = 0; j < upd_ndims; j++)
+            {
+                upd_start_offset += upd_current_indices[j] * upd_step[j];
+            }
+
+            const T* p_updates = (const T*)(updates + upd_start_offset);
+            T* p_out = (T*)(out + inp_start_offset);
+            for (int j = 0; j < updates_size; j++)
+            {
+                p_out[j] = rd(p_out[j], p_updates[j]);
+            }
```

Comment:
If no-op rd is special-cased, I think this for loop needs to be moved into those lambda expressions (rd).

Reply:
tried with `std::memcpy` instead of `std::copy`, since latter one only works with iterators and what I got are pointers not vectors. It went slower with `std::memcpy` like +100ms.

Reply:
Well that was unexpected.

Reply:
Also tested with the latest optimized impl. It is just slower (+80ms) with `std::memcpy`.